### ISSUE TITLE
Is `ManyServiceProvidersCreatedWarning` expected when adding services in OnConfiguring and "externally"

### ISSUE THREAD

**Issue by pinkfloydx33** [dotnet/efcore#30191]
This is not a bug report but rather a request for guidance

## Ask a question
Will adding extensions/configuration inside of `OnConfiguring` lead to Service Provider caching issues if you **also** add extensions via other other means (DI registration, test setup) external to the context? 

I am trying to figure out what is causing the `ManyServiceProvidersCreatedWarning` error to be emitted during our tests runs. I've read the various issues throughout the repository and I understand that it's usually indicative of SP caching not functioning due to some singleton service, etc. I've been trying to trace down the root cause and I've found a work-around that has me worried that we are configuring our contexts/applications incorrectly. I'm looking for some guidance.

I don't have a minimal repro yet as there's a lot of moving parts. In the meantime I hope that something in my description below jumps out. 

### Include your code

We have a base `DbContext` class that we share throughout all of our projects. As part of `OnConfiguring` it adds a couple of extensions--naming conventions and an internal extension (see below)--guarded behind `Options.IsFrozen`

```csharp
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
{
        base.OnConfiguring(optionsBuilder);
        if (!optionsBuilder.Options.IsFrozen)
        {
            optionsBuilder.UseSnakeCaseNamingConvention(); 
            optionsBuilder.UseOurExtensionAndConfiguration(); 
        }
}
```

In a _normal_ application we register our contexts in DI via `AddDbContext`:

```csharp
NpgsqlDataSource dbDataSource = GetDataSourceSomehow();
services.AddSomeInterceptors();
services.AddDbContext<InspectionsContext>((s, o) =>
{
        o.UseNpgsql(dbDataSource);
        o.AddInterceptors(s.GetServices<IInterceptor>());
});
```

We expect all of our applications to use the naming conventions and our custom extension. The `UseOurExtensionAndConfiguration` adds an `IDbContextOptionsExtension` which in turn registers some interceptors and a `ConventionSetPlugin` and calls `ConfigureWarnings` to set the log level for a few relational events.  We don't want to repeat it in every application which is why we've put it in our base context's `OnConfiguring` as highlighted above. So far this _seemed_ to be working fine.  

Recently, however we began to experience the `ManyServiceProvidersCreatedWarning` in our tests. We use an xUnit class fixture that initializes Postgres containers (via [Testcontainers](https://github.com/testcontainers/testcontainers-dotnet)), configures test-specific `DbContextOptions` and provides a method for our test classes to create a new context. Each class gets it's own instance of the fixture (can't be shared due to some unrelated functionality I've elided) that looks like the following:

```csharp
public abstract class TestContextFixture<T> where T : OurContextBaseClass
{
    private readonly Lazy<DbContextOptions<T>> _optionsLazy;
    protected TestContextFixture() => 
         _optionsLazy = new Lazy<DbContextOptions<T>>(CreateOptions);
    
    // filled in by Testcontainers
    protected string ConnectionString { get; set; } = null!; 
    // overridden in sub-class which passes options to context constructor
    protected abstract T CreateContext(DbContextOptions<T> options); 
    // called by Unit Tests
    public T CreateContext() => CreateContext(_optionsLazy.Value);

    private DbContextOptions<T> CreateOptions()
    {
        var options = new DbContextOptionsBuilder<T>();
        options.UseNpgsql(ConnectionString)
            .EnableDetailedErrors()
            .EnableSensitiveDataLogging()
            .ConfigureWarnings(static w =>
                w.Ignore(
                    CoreEventId.SensitiveDataLoggingEnabledWarning
                )
            );
        return options.Options;
    }
}
```

#### The workaround?
We've started to notice that as we add more test classes we are hitting the `ManyServiceProvidersCreatedWarning` issue. After a lot of guess-and-check debugging, I discovered that adding the following two lines into the `CreateOptions` method prevents this problem:

```csharp
options.UseSnakeCaseNamingConvention();
options.UseOurExtensionAndConfiguration();
```
You'll note that these are the **same two methods** that we are calling in `OnConfiguring` guarded by the `IsFrozen` check. If either  is omitted I get the errors again. It seems that EF builds the options twice--before the constructor and then again for `OnConfiguring`--considers them different, and thus must create a new internal provider. 

#### Summary/Question
Is adding extensions/performing configuring guarded behind `IsFrozen` valid, particularly if we are going to provide additional configuration "outside" the context? Is there something downright wrong about what we're doing? What's the recommended pattern for stuff like this--do I need to consolidate all extension setup into one place?

We're working on a set of greenfield applications that aren't getting much load at the moment.  I'm worried that since we don't call these methods in our _normal_ DI registration flow that we are going to eventually hit the same problem in our application as we do in our tests. If this is just likely due to the way our tests are set up I'd have no issue silencing the warning there,but I'd like to take preventative measures if this is likely a larger issue.

### Include provider and version information

EF Core version: 7.0.2
Database provider: efcore.pg
Target framework: NET7.0
Operating system: Windows11, debian (containers)
IDE: VS 2022

**Comment by ajcvickers** [comment]
> Is adding extensions/performing configuring guarded behind IsFrozen valid

Yes. However, regardless of how they are added if the configuration or extensions results in changing singleton configuration, then it will ultimately result in the problem you are seeing. It's hard to say which configuration is resulting in this without a repro.

**Comment by pinkfloydx33** [comment]
Ok that helps. Like I said there's a bunch of moving parts so a simple repro is a little difficult. Had you said "no" then case closed and I'd have saved myself some considerable effort. Given that it _should_ be valid, I will try and put together the most minimal thing I can

**Comment by pinkfloydx33** [comment]
So I was able to condense this down quite a bit. However it's still too big to fit in an issue, so I've created a repo for it: https://github.com/pinkfloydx33/ef30191

These are the lines/areas of interest: 
- [Test Setup](https://github.com/pinkfloydx33/ef30191/blob/main/Repro.Tests/TestFixture.cs#L35-L48)
- [Context](https://github.com/pinkfloydx33/ef30191/blob/main/Repro.Data/SampleContext.cs#L23-L25)
- [No-Op Extension](https://github.com/pinkfloydx33/ef30191/blob/main/Repro.Data/NoOpExtension.cs)

There are four different variants in the test-setup. The first one will throw the Many Service Providers error. The repro is configured for postgres; the connection string needs to be adapted. Now that I've stripped this down, it seems like it _could_ actually be efcore.pg specific -- FYI @roji . I don't have a way to test with Sql Server though, sorry.

**Comment by ajcvickers** [comment]
I believe the issue here is `options.UseNpgsql(_dataSource)`, which is using a new instance of DbDataSource for every fixture. I remember talking about this with @roji, but I don't remember what the guidance is.

**Comment by pinkfloydx33** [comment]
In a normal application the guidance is to use only one datasource. We can't quite do that in our tests since we are hitting different containers (ie. servers), so I think we're going to need to _eventually_ suppress the warning as we add more test classes. That should be alright for tests and is something suggested by (I think) you in one of the various issues I found on the topic. 

I guess my question becomes why does it only happen if you add the extensions from `OnConfiguring` *only*? In other words: why doesn't variant 2 in my repro produce the issue as well? It seems weird that if we re-add the extensions "on the outside" that the warning doesn't occur. Talking this out it makes sense that the pg extension can't share service providers for distinct datasources. But shouldn't EF hit that logic regardless of _where_ other extensions are added?

My main concern with the OP was that this wasn't going to rear it's head during normal application execution. Given that we use the pg features there _appropriately_, I don't think we'd ever encounter this. But given that the SP logic seems somehow dependent on _where/how_ other extensions get added I'm not 100% convinced.

**Comment by ajcvickers** [comment]
@pinkfloydx33 Yes, its seems like it should generate the warning for both cases where `options.UseNpgsql(_dataSource)` is used.

**Comment by ajcvickers** [comment]
Ping @roji

**Comment by marcwittke** [comment]
Today I fell into this pit, too.

I am heavily paralleling test runs with xUnit and create a totally separate database and DbContexts for each test (npgsql, of course). So, yes, it is totally expected to have more than 20 service providers.

workaround is:

```csharp
public static DbContextOptionsBuilder<TDbContext> UseMyDb<TDbContext>(
    this DbContextOptionsBuilder<TDbContext> dbContextBuilder,
    string connectionString,
    bool inTests = false) where TDbContext : DbContext
{
    var dataSourceBuilder = new NpgsqlDataSourceBuilder(connectionString);
    dataSourceBuilder.UseNodaTime();
    var dataSource = dataSourceBuilder.Build();

    dbContextBuilder
        .UseNpgsql(
            dataSource,
            opt =>
            {
                opt.CommandTimeout(commandTimeout: 30);
                opt.MigrationsHistoryTable("_ef_migrations_history", MyDbContext.SchemaName);
                opt.UseNodaTime();
            })
        .UseSnakeCaseNamingConvention();

    if (inTests)
    {
        // Suppress the ManyServiceProvidersCreated warning:
        // parallel test runs are expected to create a lot of service providers
        dbContextBuilder.ConfigureWarnings(builder => builder.Ignore(CoreEventId.ManyServiceProvidersCreatedWarning));
    }

    return dbContextBuilder;
}
```

I am unsure whether EF Core made a good decision in opting for guiding the noob while bothering the experienced developer.

**Comment by roji** [comment]
Note that the plan is to change EFCore.PG to make NpgsqlDataSource scoped rather than singleton (for other reasons); this would also stop requiring a new EF service provider for different data source instances, which should make this all go away.

**Comment by 0xced** [comment]
I got this `ManyServiceProvidersCreatedWarning` too but it was because I implemented a [DbContextOptionsExtensionInfo](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.infrastructure.dbcontextoptionsextensioninfo) and returned `false` for the `ShouldUseSameServiceProvider` implementation. Once I changed it to return `true` the problem vanished!
